src/runtime/HalideBuffer.h

```diff
@@ -855,8 +860,9 @@ class Buffer {
         buf.type = t;
         buf.dimensions = 1 + (int)(sizeof...(rest));
         make_shape_storage();
-        initialize_shape(0, first, rest...);
-        if (!any_zero(first, rest...)) {
+        Internal::GatheredInts<int, Args...> extents(first, rest...);
```

Comment:
Can gatheredints just be a stack array of ints?

Here's an alternative with no helpers:
```
    bool any_zero = false;
    int extents[] = {first, rest...};
    for (int e : extents) {
        any_zero |= (e == 0);
    }
    if (!any_zero) {
        int idx = 0;
        for (int e : extents) {
            buf.dim[idx].min = 0;
            buf.dim[idx].extent = e;
            idx++;
        }
    }

```

Reply:
I guess that would still imply some code duplication, but you could just stuff it in a non-recursive helper and call it like: `as<const void>().initialize_shape(first, rest...);`
to avoid duplication across template instantiations.

Reply:
I don't think `int extents[] = {first, rest...};` is legal. When I try using it I get `‘extents’ has incomplete type`.

Reply:
orly? I copied pasted that out of a test file that worked. I've been using that idiom for a while now as a simpler alternative to template-recursively gathering something. E.g. see for_each_value_impl. Here's my whole test:

```

struct d {
    int min, extent;
};
struct b {
    d dim[100];
} buf;

template<typename... Args>
void test(int first, Args... rest) {
    bool any_zero = false;
    int extents[] = {first, rest...};
    for (int e : extents) {
        any_zero |= (e == 0);
    }
    if (!any_zero) {
        int idx = 0;
        for (int e : extents) {
            buf.dim[idx].min = 0;
            buf.dim[idx].extent = e;
            idx++;
        }
    }
}

int foo(int a, int b, int c, int d) {
    test(a, b, c, d);
    return 0;
}
```
Maybe it needs an explicit type `int extents[] = {first, (int)rest...};` ?

Reply:
Hm, good thought, but `const int extents[] = {first, (int) rest...};` fails in the same way.